lte/gateway/c/oai/tasks/amf/amf_security_mode_control.cpp

```diff
@@ -0,0 +1,300 @@
+  }
+
+  amf_ue_ngap_id_t ue_id =
+      PARENT_STRUCT(amf_ctx, ue_m5gmm_context_s, amf_context)->amf_ue_ngap_id;
+  nas_amf_smc_proc_t* smc_proc = get_nas5g_common_procedure_smc(amf_ctx);
+  if (!smc_proc) {
+    smc_proc = nas5g_new_smc_procedure(amf_ctx);
+    if (smc_proc) {
+      smc_proc->saved_selected_eea =
+          amf_ctx->_security.selected_algorithms.encryption;
+      smc_proc->saved_selected_eia =
+          amf_ctx->_security.selected_algorithms.integrity;
+      smc_proc->saved_eksi     = amf_ctx->_security.eksi;
+      smc_proc->saved_overflow = amf_ctx->_security.dl_count.overflow;
+      smc_proc->saved_seq_num  = amf_ctx->_security.dl_count.seq_num;
+      smc_proc->saved_sc_type  = amf_ctx->_security.sc_type;
+      /*
+       * The security mode control procedure is initiated to take into use
+       * the 5GS  security context created after a successful execution of
+       * the 5GS authentication procedure
+       */
+      smc_ctrl.amf_ctx_set_security_eksi(amf_ctx, ksi);
+      amf_ctx->_security.dl_count.overflow = 0;
+      amf_ctx->_security.dl_count.seq_num  = 0;
+
+      rc                                                = RETURNok;
+      amf_ctx->_security.selected_algorithms.encryption = amf_eea;
+      amf_ctx->_security.selected_algorithms.integrity  = amf_eia;
+      if (rc == RETURNerror) {
+        OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNerror);
+      }
+      smc_ctrl.amf_ctx_set_security_type(
+          amf_ctx, SECURITY_CTX_TYPE_FULL_NATIVE);
+      derive_key_nas(
+          NAS_INT_ALG, amf_ctx->_security.selected_algorithms.integrity,
+          amf_ctx->_vector[amf_ctx->_security.eksi % MAX_EPS_AUTH_VECTORS]
+              .kasme,
+          amf_ctx->_security.knas_int);
+      derive_key_nas(
+          NAS_ENC_ALG, amf_ctx->_security.selected_algorithms.encryption,
+          amf_ctx->_vector[amf_ctx->_security.eksi % MAX_EPS_AUTH_VECTORS]
+              .kasme,
+          amf_ctx->_security.knas_enc);
+      /*
+       * Set new security context indicator
+       */
+      security_context_is_new = true;
+      amf_ctx_set_attribute_present(amf_ctx, AMF_CTXT_MEMBER_SECURITY);
+    }
+  } else {
+    OAILOG_ERROR(LOG_NAS_AMF, "AMF-PROC  - No 5G CN security context exists\n");
+    OAILOG_FUNC_RETURN(LOG_NAS_AMF, RETURNerror);
+  }
+
+  if (smc_proc) {
+    /*
+     * Setup ongoing AMF procedure callback functions
+     */
+    smc_proc->amf_com_proc.amf_proc.delivered               = NULL;
+    smc_proc->amf_com_proc.amf_proc.base_proc.success_notif = success;
+    smc_proc->amf_com_proc.amf_proc.base_proc.failure_notif = failure;
+    smc_proc->amf_com_proc.amf_proc.base_proc.fail_in       = NULL;
+    smc_proc->amf_com_proc.amf_proc.base_proc.fail_out      = NULL;
+    /*
+     * Set the UE identifier
+     */
+    smc_proc->ue_id = ue_id;
+    /*
+     * Reset the retransmission counter
+     */
+    smc_proc->retransmission_count = 0;
+    /*
+     * Set the key set identifier
+     */

```

Comment:
Remove all bloated comments, these are unnecessary for simple statements

Reply:
Taken care